system-helper/flatpak-system-helper.c

```diff
@@ -190,6 +420,341 @@ get_sender_pid (GDBusMethodInvocation *invocation)
+      g_autofree gchar *child_repo_path = NULL;
+      g_autofree gchar *user_dir = NULL;
+      g_autofree gchar *root_dir = NULL;
+      g_autofree gchar *flatpak_dir = NULL;
+      g_autofree gchar *arg_user = NULL;
+      g_autofree gchar *out_location = NULL;
+      guint uid_pull_count = 0;
+      uid_t uid;
+
+      if (!get_connection_uid (invocation, &uid, &error))
+        {
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return TRUE;
+        }
+
+      uid_pull_count = get_uid_pull_count (uid);
+      if (uid_pull_count > MAX_PULLS_PER_USER)
+        {
+          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_LIMITS_EXCEEDED,
+                                                 "Maximum number of pulls per user exceeded for UID: %u", uid);
+          return TRUE;
+        }
+
+      flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+      g_autofree gchar *repo_tmp = g_build_filename (flatpak_dir, "repo", "tmp", NULL);
+
+      g_auto(GStrv) parts = flatpak_decompose_ref (ref, NULL);
+      if (check_partial_pull_availability (repo_tmp, parts[1], commit, &out_location))
+        {
+          g_autoptr(GFile) user_dir_file = NULL;
+
+          g_debug ("Found a partial pull for ref %s at: %s", parts[1], out_location);
+          root_dir = g_path_get_dirname (out_location);
+          user_dir = g_strdup_printf ("%s-bindfs", root_dir);
+          user_dir_file = g_file_new_for_path (user_dir);
+          if (!g_file_query_exists (user_dir_file, NULL))
+            g_mkdir_with_parents (user_dir, 0755);
+          child_repo_path = g_steal_pointer (&out_location);
+        }
+      else
+        {
+          g_autoptr(OstreeRepo) child_repo = NULL;
+          g_auto(GLnxLockFile) child_repo_lock = { 0, };
+
+          g_autofree gchar *cache_dir_basename = g_strdup_printf ("flatpak-cache-%s-XXXXXX", parts[1]);
+          root_dir = g_mkdtemp_full (g_build_filename (repo_tmp, cache_dir_basename, NULL), 0755);
+          user_dir = g_strdup_printf ("%s-bindfs", root_dir);
+          if (g_mkdir_with_parents (user_dir, 0755) == -1)
+            {
+              glnx_throw_errno_prefix (&error, "Can't create directory for bindfs mount at %s", user_dir);
+              g_dbus_method_invocation_return_gerror (invocation, error);
+              return TRUE;
+            }
+
+          g_autoptr(GFile) root_dir_file = g_file_new_for_path (root_dir);
+          child_repo = flatpak_dir_create_child_repo (system, root_dir_file, &child_repo_lock, NULL, &error);
+          if (child_repo == NULL)
+            {
+              g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                                     "Failed to create OSTree child repo at %s", root_dir);
+              return TRUE;
+            }
+          GFile *repo_file = ostree_repo_get_path (child_repo);
```

Comment:
this one looks pointless; can just do

child_repo_path = g_file_get_path (ostree_repo_get_path (child_repo));

Reply:
not what I meant, though. I'd like to get rid of the local altogether